domain/machine/state/machine_cloud_instance.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/instance"
+	"github.com/juju/juju/domain"
+)
+
+// HardwareCharacteristics returns the hardware characteristics struct with
+// data retrieved from the machine_cloud_instance table.
+func (st *State) HardwareCharacteristics(
+	ctx context.Context,
+	machineUUID string,
+) (*instance.HardwareCharacteristics, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	retrieveHardwareCharacteristics := `
+SELECT (*) AS (&instanceData.*)
+FROM   machine_cloud_instance 
+WHERE  machine_uuid = $instanceData.machine_uuid`
+	machineUUIDQuery := instanceData{MachineUUID: machineUUID}
+	retrieveHardwareCharacteristicsStmt, err := st.Prepare(retrieveHardwareCharacteristics, machineUUIDQuery, sqlair.M{})
```

Comment:
What's the `sqlair.M` for?

Reply:
No use whatsoever, that's why I removed it in 29982a633a :)